File: client/python/easemlclient/easemlclient/model/job.py
"""
Implementation of the `Job` class.
"""
import pyrfc3339

from copy import deepcopy
from datetime import datetime, timedelta
from enum import Enum
from typing import Dict, Optional, Any, Iterator, Tuple, List

from .core import Connection
from .common import TimeInterval
from .dataset import Dataset
from .module import Module
from .process import Process
from .user import User
from .type import ApiType, ApiQuery, ApiQueryOrder


class JobStatus(Enum):
    SCHEDULED = "scheduled"
    RUNNING = "running"
    PAUSING = "pausing"
    PAUSED = "paused"
    RESUMING = "resuming"
    COMPLETED = "completed"
    TERMINATING = "terminating"
    TERMINATED = "terminated"
    ERROR = "error"


class Job(ApiType['Job']):
    """The Job class contains information about datasets.

    ...
    Attributes:
    -----------
    identifier: str
        A unique identifier of the user (i.e. the username).
    name: str
        The full name of the user.
    status: str
        The current status of the user. Can be 'active' or 'archived'.
    """

    def __init__(self, input: Dict[str, Any]) -> None:
        if "id" not in input:
            raise ValueError("Invalid input dictionary: It must contain an 'id' key.")

        super().__init__(input)

    @classmethod
    def create(cls, dataset: Dataset, objective: Module, models: List[Module],
               accept_new_models: bool = True, max_tasks: int = 100,
               alt_objectives: Optional[List[Module]] = None, config_space: Optional[Dict[str, Any]] = None) -> 'Job':
        init_dict: Dict[str, Any] = {"id": None}
        if dataset is not None:
            init_dict["dataset"] = dataset.id
        if objective is not None:
            init_dict["objective"] = objective.id
        if models is not None:
            init_dict["models"] = [x.id for x in models]
        if accept_new_models is not None:
            init_dict["accept-new-models"] = accept_new_models
        if max_tasks is not None:
            init_dict["max-tasks"] = max_tasks
        if alt_objectives is not None:
            init_dict["alt-objectives"] = [x.id for x in alt_objectives]
        if config_space is not None:
            init_dict["config-space"] = config_space
        return Job(init_dict)
    
    @classmethod
    def create_ref(cls, id: str) -> 'Job':
        return Job({"id": id})

    @property
    def id(self) -> str:
        return self._dict["id"]

    @property
    def user(self) -> Optional[User]:
        value = self._dict.get("user")
        return User({"id": value}) if value is not None else None

    @property
    def dataset(self) -> Optional[Dataset]:
        value = self._dict.get("dataset")
        return Dataset({"id": value}) if value is not None else None

    @property
    def models(self) -> Optional[List[Module]]:
        value = self._dict.get("models")
        return [Module({"id": x}) for x in value] if value is not None else None

    @property
    def config_space(self) -> Optional[str]:
        value = self._dict.get("config-space")
        return str(value) if value is not None else None

    @property
    def accept_new_models(self) -> Optional[bool]:
        value = self._updates.get("accept-new-models") or self._dict.get("accept-new-models")
        return bool(value) if value is not None else None

    @accept_new_models.setter
    def accept_new_models(self, value: Optional[bool] = None) -> None:
        if value is not None:
            self._updates["accept-new-models"] = value
        else:
            self._updates.pop("accept-new-models")

    @property
    def objective(self) -> Optional[Module]:
        value = self._dict.get("objective")
        return Module({"id": value}) if value is not None else None

    @property
    def alt_objectives(self) -> Optional[List[Module]]:
        value = self._dict.get("alt-objectives")
        return [Module({"id": x}) for x in value] if value is not None else None

    @property
    def max_tasks(self) -> Optional[int]:
        value = self._updates.get("max-tasks") or self._dict.get("max-tasks")
        return int(value) if value is not None else None

    @max_tasks.setter
    def max_tasks(self, value: Optional[int] = None) -> None:
        if value is not None:
            self._updates["max-tasks"] = value
        else:
            self._updates.pop("max-tasks")

    @property
    def creation_time(self) -> Optional[datetime]:
        value = self._dict.get("creation-time")
        return pyrfc3339.parse(value) if value is not None else None

    @property
    def running_time(self) -> Optional[TimeInterval]:
        value = self._dict.get("running-time")
        return TimeInterval(value) if value is not None else None

    @property
    def running_duration(self) -> Optional[timedelta]:
        value = self._dict.get("running-duration")
        return timedelta(milliseconds=int(value)) if value is not None else None

    @property
    def pause_duration(self) -> Optional[timedelta]:
        value = self._dict.get("pause-duration")
        return timedelta(milliseconds=int(value)) if value is not None else None

    @property
    def status(self) -> Optional[JobStatus]:
        value = self._updates.get("status") or self._dict.get("status")
        return JobStatus(value) if value is not None else None

    @status.setter
    def status(self, value: Optional[JobStatus] = None) -> None:
        if value is not None:
            self._updates["status"] = value.value
        else:
            self._updates.pop("status")

    @property
    def status_message(self) -> Optional[str]:
        value = self._dict.get("status-message")
        return str(value) if value is not None else None

    @property
    def process(self) -> Optional[Process]:
        value = self._dict.get("process")
        return Process({"id": value}) if value is not None else None

    def __iter__(self) -> Iterator[Tuple[str, Any]]:
        for (k, v) in self._dict:
            yield (k, v)

    def post(self, connection: Connection) -> 'Job':
        url = connection.url("jobs")
        return self._post(connection, url)

    def patch(self, connection: Connection) -> 'Job':
        url = connection.url("jobs/" + self.id)
        return self._patch(connection, url)

    def get(self, connection: Connection) -> 'Job':
        url = connection.url("jobs/" + self.id)
        return self._get(connection, url)

class JobQuery(ApiQuery['Job', 'JobQuery']):

    VALID_SORTING_FIELDS = ["user", "dataset", "objective", "creation-time", "running-time-start", "running-time-end", "status"]

    def __init__(self, id: Optional[List[str]] = None, user: Optional[User] = None,
                 dataset: Optional[Dataset] = None, model: Optional[Module] = None,
                 objective: Optional[Module] = None, alt_objective: Optional[Module] = None,
                 status: Optional[JobStatus] = None, accept_new_models: Optional[bool] = None,                
                 order_by: Optional[str] = None, order: Optional[ApiQueryOrder] = None,
                 limit: Optional[int] = None, cursor: Optional[str] = None) -> None:
        super().__init__(order_by, order, limit, cursor)
        self.T = Job

        if id is not None:
            self._query["id"] = id
        if user is not None:
            self._query["user"] = user.id
        if dataset is not None:
            self._query["dataset"] = dataset.id
        if model is not None:
            self._query["model"] = model.id
        if objective is not None:
            self._query["objective"] = objective.id
        if alt_objective is not None:
            self._query["alt-objective"] = alt_objective.id
        if status is not None:
            self._query["status"] = status.value
        if accept_new_models is not None:
            self._query["accept-new-models"] = accept_new_models

    def run(self, connection: Connection) -> Tuple[List[Job], Optional['JobQuery']]:
        url = connection.url("jobs")
        return self._run(connection, url)

File: engine/database/model/types/job.go
package types

import (
	"time"

	"github.com/globalsign/mgo/bson"
)

const (
	// JobScheduled is a job that is scheduled but not running yet.
	JobScheduled = "scheduled"

	// JobRunning is a job that is picked up by the scheduler and started running.
	JobRunning = "running"

	// JobPausing is a job that is in a pausing state. All its tasks are being paused.
	JobPausing = "pausing"

	// JobPaused is a job that is paused but may be resumed.
	JobPaused = "paused"

	// JobResuming is a job that is leaving the paused state. All its paused tasks are being resumed.
	JobResuming = "resuming"

	// JobCompleted is a job that has been completed.
	JobCompleted = "completed"

	// JobTerminating is a job that is in a terminating state. All its tasks are being terminated.
	JobTerminating = "terminating"

	// JobTerminated is a job that was terminated before completion.
	JobTerminated = "terminated"

	// JobError is a job that is in an error state. The error information is logged.
	JobError = "error"

	// DefaultMaxTasks is the default number of tasks per job.
	DefaultMaxTasks = 100
)

// Job contains information about jobs.
type Job struct {
	ID                bson.ObjectId `bson:"_id" json:"id"`
	User              string        `bson:"user" json:"user"`
	Pipeline   		  []string      `bson:"pipeline" json:"pipeline"`
	TaskIds	   		  []string      `bson:"task-ids" json:"task-ids"`
	Dataset           string        `bson:"dataset" json:"dataset"`
	Models            []string      `bson:"models" json:"models"`
	ConfigSpace       string        `bson:"config-space" json:"config-space"`
	AcceptNewModels   bool          `bson:"accept-new-models" json:"accept-new-models"`
	Objective         string        `bson:"objective" json:"objective"`
	AltObjectives     []string      `bson:"alt-objectives" json:"alt-objectives"`
	MaxTasks          uint64        `bson:"max-tasks" json:"max-tasks"`
	CreationTime      time.Time     `bson:"creation-time" json:"creation-time"`
	RunningTime       TimeInterval  `bson:"running-time" json:"running-time"`
	RunningDuration   uint64        `bson:"running-duration,omitempty" json:"running-duration"`
	PauseStartTime    time.Time     `bson:"pause-start-time"`
	PauseDuration     uint64        `bson:"pause-duration,omitempty" json:"pause-duration"`
	PrevPauseDuration uint64        `bson:"prev-pause-duration"`
	Status            string        `bson:"status" json:"status"`
	StatusMessage     string        `bson:"status-message" json:"status-message"`
	Process           bson.ObjectId `bson:"process,omitempty" json:"process"`
}

// IsStarted returns true when the job has passed the "scheduled" state.
func (job Job) IsStarted() bool {
	return job.Status != JobScheduled
}

// IsPaused returns true when the job is in the paused state.
func (job Job) IsPaused() bool {
	return job.Status == JobPaused
}

// IsEnded returns true when the job has either completed, terminated or is in an error state.
func (job Job) IsEnded() bool {
	return job.Status == JobCompleted || job.Status == JobTerminated || job.Status == JobError
}

// GetPauseDuration computes the total time that the job has spent in the paused state.
func (job Job) GetPauseDuration() uint64 {
	pauseDuration := job.PrevPauseDuration
	if job.IsPaused() {
		pauseDuration += uint64(time.Since(job.PauseStartTime).Nanoseconds() / 1000000)
	}
	return pauseDuration
}

// GetRunningDuration computes the total time the job has spent running (including pauses).
func (job Job) GetRunningDuration() uint64 {
	var runningDuration uint64
	if job.IsEnded() {
		runningDuration = uint64(job.RunningTime.End.Sub(job.RunningTime.Start).Nanoseconds() / 1000000)

	} else if job.IsStarted() {
		runningDuration = uint64(time.Since(job.RunningTime.Start).Nanoseconds() / 1000000)
	}
	return runningDuration
}
